Add Terraformer instance type and other modifications for RTA

jsf9k opened the pull request:
## 🗣 Description ##

This pull request adds Terraformer instances to this repository and makes assorted other changes for RTA.

## 💭 Motivation and context ##

See cisagov/cool-system-internal#5.

## 🧪 Testing ##

I used these changes to successfully spin up a Teamserver (using the Terraformer role created herein).  All changes are already applied to env3 in our staging COOL environment.

## ✅ Checklist ##

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] All new and existing tests pass.

jsf9k committed:
Fix a typo

jsf9k committed:
Add a DynamoDB VPC gateway endpoint

Terraformer instances will want to use this endpoint to acquire locks
before modifying the remote state.

jsf9k committed:
Add terraformer instance type to variables

jsf9k committed:
Add a security group for terraformer instances

jsf9k committed:
Add TF code for the terraformer instance itself

jsf9k committed:
Add Terraformer role and allow Terraformer instances to assume it

jsf9k committed:
Add cloud-init code to create an AWS CLI configuration file

The file is created in /var/terraformer/aws_config and can be utilized
by a user by simply copying it to ~/.aws/config.

jsf9k committed:
Give terraformer SG access to interface VPC endpoints

jsf9k committed:
Attach the SSM Agent policy to the terraformer instance role

jsf9k committed:
Put Guacamole in the same subnet as Guacamole for now

For some reason I can't SSM to the instance unless I put it in the
first private subnet.  I believe this has something to do with the
NACLs that are in place for that subnet specifically.  I will figure
this out later.  I'd definitely prefer to put this instance in a
different subnet than the Guacamole instance.

jsf9k committed:
Create the AWS config file directly in the VNC user's home dir

jsf9k committed:
Adjust permissions and ownership of AWS config

jsf9k committed:
Give the Terraformer role the correct assume role policy

jsf9k committed:
Enforce the use of STS regional endpoints

This is required when using VPC interface endpoints to access STS.

jsf9k committed:
Move comment

jsf9k committed:
Add some bash to ensure that /home/vnc/.aws is owned by vnc

jsf9k committed:
Refine Terraformer instance role permissions

Now the instance role is granted read-only permission to everything
but is only allowed to modify or destroy resources that are not tagged
as belonging to the "VM Fusion - Development" team.

jsf9k committed:
Add the Terraformer instances to the Guacamole connections

jsf9k committed:
Update README.md via terraform-docs

jsf9k committed:
Revert "Remove SG and ACL rules related to the (now-removed) NAT gateway"

This reverts commit 776c9f63d11a43ab0a6a3b731b90f2422056ab99.

jsf9k committed:
Revert "Remove the NAT gateway and associated route"

This reverts commit ed3c3d0796d6750b80901f7ddacd6d98c92aae43.

dav3r commented on terraformer_cloud_init.tf:
```diff
@@ -0,0 +1,30 @@
+    content = templatefile(
+      "${path.module}/cloud-init/write-aws-config.tpl.yml", {
+        aws_region  = var.aws_region
+        owner       = "vnc:vnc"
+        path        = "/home/vnc/.aws/config"
```
The VNC user's username should be pulled in from the SSM parameter store (see `var.ssm_key_vnc_username`).

jsf9k replied:
See commit f146058.

dav3r replied:
That's the stuff right there! 🪐

dav3r commented on terraformer_ec2.tf:
```diff
@@ -0,0 +1,72 @@
+  availability_zone    = "${var.aws_region}${var.aws_availability_zone}"
+  iam_instance_profile = aws_iam_instance_profile.terraformer.name
+  instance_type        = "t3.medium"
+  # For some reason I can't SSM to the instance unless I put it in the
+  # first private subnet.  I believe this has something to do with the
+  # NACLs that are in place for that subnet specifically.  I will
+  # figure this out later.  I'd definitely prefer to put this instance
+  # in a different subnet than the Guacamole instance.
```
It sounds like this should be a `TODO` with a new issue to go along with it.

jsf9k replied:
See commit 5bc0ddb.

jsf9k committed:
Remove explicit tags from resources re-added by revert

These are no longer necessary since we use default_tags on the
providers.

jsf9k committed:
Remove HTTPS egress from Guacamole SG

This was re-added by the revert but this egress is no longer needed
because the required Docker image is now baked into the AMI.

jsf9k committed:
Add a rule to allow the private subnets to ingress via port 80

This is necessary so that they can egress to the internet via port 80
to install packages.

Also fix the comments for the HTTPS ingress and ephemeral port egress
ACL rules.  These comments used to describe how the rules were
necessary for the Guacamole instance to download its Docker image, but
they are now only necessary for the Terraform instance to install
packages and perform terraform init.

dav3r commented on terraformer_policy.tf:
```diff
@@ -0,0 +1,34 @@
+  provider = aws.provisionassessment
+
+  # Allow full access to new resources and existing resources that
+  # _were not_ created by the "VM Fusion - Development" team.
```
Shouldn't this comment be written more generically, since we are comparing against the value of `var.tags["Team"]` (which may or may not be "VM Fusion - Development")?

jsf9k replied:
See commit acde207.

dav3r replied:
That works for me, thanks! 👍

jsf9k committed:
Add a rule allowing egress anywhere via HTTP

This is necessary for the Terraformer instance to update and install
packages.

Also fix a comment on the HTTPS egress rule that mentions it being
necessary for the Guacamole instance to download its Docker image.
This is no longer the case, but it _is_ necessary for the Terraformer
instances to perform terraform init.

jsf9k committed:
Add SG rules to allow Terraformer instances HTTP and HTTPS egress everywhere

This is necessary for these instances to install/update packages and
perform terraform init, respectively.

dav3r commented on terraformer_role.tf:
```diff
@@ -0,0 +1,28 @@
+  role       = aws_iam_role.terraformer_role.name
+}
+
+# Grant full access to anything that was not created by the "VM Fusion
+# - Development" team
```
As above, this comment should be more generic.

jsf9k replied:
See commit acde207.

dav3r reviewed (approved):
This looks excellent to me, though I did note a few small items for you to take a look at. 🏗️

jsf9k committed:
Allow Terraformer instances to egress on port 22 (ssh)

This is necessary so they can configure redirector instances via
Ansible.

jsf9k committed:
Add a read-only role for this root module's Terraform state

This will be useful for the Terraformer instance for extracting subnet
ARNs, subnet CIDRs, etc.

jsf9k committed:
Configure the Terraformer instance to be able to assume the read-only TF state role via an AWS profile

Also move the AWS configuration file to ~/.aws/credentials since
profiles are not picked up by AWS CLI when it is located in
~/.aws/config.

jsf9k committed:
Add subnets and relevant security groups to the outputs

jsf9k committed:
Add EFS client SG and EFS mount targets to outputs

jsf9k committed:
Add assessment private DNS zone to outputs

jsf9k committed:
Add email-sending domain cert read role to outputs

jsf9k committed:
Do not create the terraform state read role assume role

This makes use of the new create_assume_role variable from
cisagov/terraform-state-read-role-tf-module#7.

jsf9k committed:
Revert to using the default branch of the cisagov/terraform-state-read-role-tf-module

The branch we were using was merged as part of
cisagov/terraform-state-read-role-tf-module#7.

jsf9k committed:
Allow the Terraformer instance role to assume the org read-only role

jsf9k committed:
Add the org read role to the Terraformwe AWS config created by cloud-init

Also make the profile names more specific.

jsf9k committed:
Add permissions to the Terraformer role to allow launching of COOL instance types

Also rename some resources to better reflect the new permissions added
to the policy.

jsf9k committed:
Remove count where unnecessary on Nessus resources

I noticed that we were creating identical policies and roles for each
Nessus instance, but then using only the zeroth one, so this change
gets rid of some duplicate resources.  It also makes it much easier
for me to make the Nessus instance role an output so that it can be
used by the Terraformer instance.

jsf9k committed:
Add instance profiles as outputs

jsf9k committed:
Update README via terraform-docs

jsf9k committed:
Add more outputs

jsf9k committed:
Make a few comments more general

Co-authored-by: Dave Redmin <david.redmin@beta.dhs.gov>

jsf9k committed:
Rework write-aws-config.tpl.yml as a bash script

This will make it possible to pull the VNC user's name from SSM Parameter Store.

jsf9k committed:
Give the Terraformer instance role permission to assume the VNC Parameter Store read-only role

This is so the cloud-init script that writes out the AWS config file
can assume this role to read the VNC user's username.

jsf9k committed:
Pull the VNC username from SSM Parameter Store when creating the AWS config

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Turn comment into a TODO with a link to a Github issue

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on outputs.tf:
```diff
@@ -1,14 +1,149 @@
+
+output "debian_desktop_instance_profile" {
+  value       = aws_iam_instance_profile.debiandesktop
+  description = "The instance profile for the Debian Desktop instances."
```
I think we have standardized on the capitalization below in this this repo:

```suggestion
  description = "The instance profile for the Debian desktop instances."
```

dav3r commented on terraformer_iam.tf:
```diff
@@ -0,0 +1,64 @@
+# Define the role policies below
+################################
+
+# Allow the Terraformer instance to assume the necessary role
```
Missing a word:
```suggestion
# Allow the Terraformer instance to assume the necessary role to
```

jsf9k committed:
Add a missing word to a comment

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Change capitalization in the interest of consistency

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
Outstanding work on this PR - everything looks great! 🏋️

jsf9k committed:
Add example code showing how to make use of a Terraformer instance

jsf9k committed:
Configure the mypy pre-commit hook to skip a directory in the example code

At least one file in this directory is a duplicate of a file in
<root_dir>/cloud-init, which confuses mypy, we skip it.

jsf9k committed:
Add the terraformer SG and the assessment VPC to the outputs

jsf9k committed:
Merge branch 'develop' into improvement/add-terraformer-instance

jsf9k committed:
Allow the Terraformer instance to create new security groups

jsf9k committed:
Add a permission that allows SGs to be added and removed from instances

jsf9k committed:
Add assessment instances to outputs

jsf9k committed:
Add the NAT GW for the provate subnets as an output

dav3r reviewed (approved):
Still looks great! 🦚 

The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

jsf9k commented:
> Still looks great! 
> 
> The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

When you say "aliasing" do you mean "symlinking"?

dav3r commented:
> > Still looks great!
> > The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).
> 
> When you say "aliasing" do you mean "symlinking"?

Yes, that is indeed what I meant.

jsf9k committed:
Change duplicate files in example to symlinks

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k commented:
> Still looks great! 
> 
> The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).

Check out commit c4e1c1a.

dav3r commented:
> > Still looks great!
> > The only thought I had was whether it's possible to reduce duplication by aliasing some of the example files that are identical to other files in this repo (e.g. the cloud-init scripts).
> 
> Check out commit [c4e1c1a](https://github.com/cisagov/cool-assessment-terraform/commit/c4e1c1ad273c2064da175c2f263d92638a7b95b7).

Lovely! 🚀

jsf9k committed:
Add role to Terraformer instances to allow Route53 access to the RTA redirector account

dav3r commented on cloud-init/write-aws-config.tpl.sh:
```diff
@@ -0,0 +1,107 @@
+# * organization_read_role_arn - the ARN of the IAM role that can be
+#   assumed to read information about the AWS Organization to which
+#   the assessment environment belongs
+# * redirector_route53_role_arn - the ARN of the IAM role that can be
+#   assumed to create, delete, and modify Route53 records in the RTA
+#   redirector account
```
Is it worth generalizing this variable so that it is not so explicitly tied to RTA and their redirector account?  The answer may be "no", but I wanted to make sure we at least discussed it.  I don't have a great name suggestion, but I was thinking something along the lines of `assessor_owned_route53_role_arn` - that can be a starting point - I'm sure we can do better.

jsf9k replied:
Really I'm just creating the profile.  If you have a role from any other account for which you have delegated authority to this assessment account then you could put it there.  The RTA folks just happen to need to give us a role to allow them to make Route53 changes in their own account.  In that sense the profile is very general and I could just as well call it `role_arn`.  We won't know or care what they use the profile to do.

That said, I hadn't really envisioned any other uses for this role.  I guess I'm fine with changing the name if you and @mcdonnnj can decide on what you want.

dav3r replied:
The general nature of this role makes it tough to name, but I suppose as long as we document it well enough then that won't matter.  I'd be fine with explicitly referencing this RTA/Route53 use case in the variable description.  What do you think about `assessment_team_delegated_account_role_arn`?

jsf9k replied:
I ended up going with `assessor_account_role_arn` as I felt the rest was redundant.  See commit 65b1e21.

dav3r replied:
`assessor_account_role_arn` works for me - thanks!

jsf9k committed:
Add the operations and private subnet ACLs to the outputs

jsf9k committed:
Allow terraformer instances to modify network ACL entries

This is for the operations subnet only.

jsf9k committed:
Rename input variable to something more generic

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on README.md:
```diff
@@ -346,35 +384,39 @@ the COOL environment.
 | Name | Description | Type | Default | Required |
 |------|-------------|------|---------|:--------:|
 | assessment\_account\_name | The name of the AWS account for this assessment (e.g. "env0"). | `string` | n/a | yes |
+| assessor\_role\_role\_arn | The ARN of an IAM role that can be assumed to create, delete, and modify AWS resources in a separate assessor-owned AWS account. | `string` | `"arn:aws:iam::123456789012:role/Allow_It"` | no |
```
You mixed up the variable name here:
```suggestion
| assessor\_account\_role\_arn | The ARN of an IAM role that can be assumed to create, delete, and modify AWS resources in a separate assessor-owned AWS account. | `string` | `"arn:aws:iam::123456789012:role/Allow_It"` | no |
```

dav3r reviewed (approved):
Aside from the small README item I just mentioned, this all looks great! 🐲

jsf9k committed:
Fix incorrect variable name

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

mcdonnnj commented on terraformer_sg.tf:
```diff
@@ -0,0 +1,98 @@
+
+# Allow egress anywhere via HTTPS
+#
+# For: Terraformer instances must be able terraform init.
```
```suggestion
# For: Terraformer instances must be able to terraform init.
```

jsf9k committed:
Merge remote-tracking branch 'origin/lineage/skeleton' into improvement/add-terraformer-instance

jsf9k committed:
Point to the cloud-init scripts in the parent project

Local symlinks are not sufficient to pass the terraform-validate
pre-commit hook.

dav3r reviewed (approved):
👍 👍

jsf9k committed:
Add the COOL VPN CIDR block as an output

The unnamed assessors require this information for unspecified
purposes.

jsf9k committed:
Allow Terraformer instances to deploy instances to the private subnet

jsf9k committed:
Add ACLs to allow private subnet to access port 5986 in the operations subnet

Port 5986 is used for Windows Remote Management, which Ansible uses
when configuring Windows instances.

I also took the opportunity to renumber a lot of ACL rules.  I have
finally grown tired of having to renumber a bunch rules just to insert
one or two.

jsf9k committed:
Allow Terraformer instances to access port 5986 in the operations subnet

This port is used by Windows Remote Management, which is what Ansible
uses to configure Windows instances.

dav3r commented on .github/workflows/build.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
     runs-on: ubuntu-latest
     steps:
-      - uses: cisagov/setup-env-github-action@develop
+      - uses: cisagov/setup-env-github-action@improvement/support_tf_0.13
```
This change is reverted in https://github.com/cisagov/cool-assessment-terraform/pull/136, but just wanted to mention it to make sure that it flows into this PR.

jsf9k replied:
Yep.  Taken care of in commit 1110e86.

dav3r commented on terraformer_sg.tf:
```diff
@@ -0,0 +1,113 @@
+  to_port           = 22
+}
+
+# Allow egress anywhere via port 5986 (Windows Remote Manager).
+#
+# For: Terraformer instances must be able to configure Windows-based
+# operations instances via Ansible.
+resource "aws_security_group_rule" "terraformer_egress_to_operations_via_winrm" {
+  provider = aws.provisionassessment
+
+  security_group_id = aws_security_group.terraformer.id
+  type              = "egress"
+  protocol          = "tcp"
+  cidr_blocks       = [aws_subnet.operations.cidr_block]
+  from_port         = 5986
+  to_port           = 5986
+}
```
Is there a reason this block is not at the bottom of this file?  It breaks the order of our rules being sorted by port number.

jsf9k replied:
I was thinking that the `ssh` and `winrm` jazz should live together since it serves the same purpose.  You're argument is just as valid, though, so I'll move it.

jsf9k replied:
See commit 200ce1b.

dav3r replied:
I knew there was some method to your madness, but I wasn't sure what it was.  Thanks for moving that block. 👍

dav3r reviewed (approved):
Approved, with a couple of comments. 👍

jsf9k committed:
Move the winrm rule to the bottom of the file

@dav3r pointed out that this way the SG rules are in order of
increasing egress port number.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

jsf9k committed:
Merge branch 'develop' into improvement/add-terraformer-instance

mcdonnnj commented on .pre-commit-config.yaml:
```diff
@@ -98,6 +98,10 @@ repos:
+        # At least one file in this directory is a duplicate of a file
+        # in <root_dir>/cloud-init, which confuses mypy, so let's skip
+        # it.
+        exclude: examples/use-terraformer-instance/cloud-init
```
I no longer see this directory. Is this still needed?

```suggestion
```

mcdonnnj commented on cloud-init/write-aws-config.tpl.sh:
```diff
@@ -0,0 +1,107 @@
+# templating engine balks at that; hence, I am using dirname instead.
+d=$(dirname "$path")
+mkdir -p "$d"
+chown -R "$vnc_username:$vnc_username" "$d"
```
Are we unable to use the long flags here?

```suggestion
mkdir --parents "$d"
chown --recursive "$vnc_username:$vnc_username" "$d"
```

jsf9k committed:
Allow VPN users to access more than just Guacamole

This involves:
* Adding a new local variable to enumerate the ports required for
access
* Opening some more ports in the private subnet ACLs

Some assessors want to host things like Mattermost alongside Guacamole
in their assessment environments.

jsf9k commented:
The checks should pass again once the dust settles on our Terraform upgrade.  I will re-run the checks once we're in a good state there.  I'm also waiting until then to attempt to apply the changes from the most recent commit to an environment in our staging COOL environment.

mcdonnnj commented on outputs.tf:
```diff
@@ -1,14 +1,219 @@
+  description = "The Assessor Portal instances."
+}
+
+output "assessorportal_security_group" {
```
Is there a reason this output is named differently from the previous two for the same group of resources?

```suggestion
output "assessor_portal_security_group" {
```

jsf9k replied:
See commit b7a7c97.

mcdonnnj commented on outputs.tf:
```diff
@@ -1,14 +1,219 @@
+
 output "guacamole_server" {
   value       = aws_instance.guacamole
   description = "The AWS EC2 instance hosting guacamole."
```
Should this be a proper noun?

```suggestion
  description = "The AWS EC2 instance hosting Guacamole."
```

jsf9k replied:
Please see commit a22bb6456830b2526ec2ff47b45aa94ba1435e80.

mcdonnnj commented on outputs.tf:
```diff
@@ -1,14 +1,219 @@
+  description = "The access control list (ACL) for the operations subnet."
+}
+
+output "pentestportal_instance_profile" {
```
Should these all be prefixed `pentest_portal_*`? That would be in line with 2/3 `assessor_portal_*` outputs and all `debian_desktop_*` outputs previously declared.

jsf9k replied:
Please see commit fc8e68b.

mcdonnnj commented on private_acl_rules.tf:
```diff
@@ -1,26 +1,84 @@
-# For: Assessment team access to guacamole web client
-resource "aws_network_acl_rule" "private_ingress_from_cool_vpn_via_https" {
+# For: Assessment team access to services hosted in the private subnet
+# (i.e. guacamole, Mattermost, etc.)
```
```suggestion
# (i.e. Guacamole, Mattermost, etc.)
```

mcdonnnj commented on private_acl_rules.tf:
```diff
@@ -1,26 +1,84 @@
+# (i.e. guacamole, Mattermost, etc.)
+resource "aws_network_acl_rule" "private_ingress_from_cool_vpn" {
+  provider = aws.provisionassessment
+  for_each = [
```
It might just be me but there's enough going on here that a brief comment might be good for future skimmers.

jsf9k replied:
Please see commit 68bf56a.

mcdonnnj commented on providers.tf:
```diff
@@ -88,3 +88,18 @@ provider "aws" {
+
+# This provider is required by the read_terraform_state module in
+# read_terraform_state_role.tf in order to create the read-only role
+# for this Terraform root module's terraform state.
```
```suggestion
# for this Terraform root module's Terraform state.
```

mcdonnnj commented on read_terraform_state_role.tf:
```diff
@@ -0,0 +1,31 @@
+    # to create IAM resources in the account(s) listed in account_ids.
+    # In any event, this provider is not used here at all since the
+    # module is configured with create_assume_role equal to false.  We
+    # do have to put something there are a placeholder, though.
```
```suggestion
    # do have to put something there as a placeholder, though.
```

mcdonnnj commented on terraformer_assume_role_policy_doc.tf:
```diff
@@ -0,0 +1,19 @@
+data "aws_iam_policy_document" "terraformer_assume_role_doc" {
+  statement {
+    actions = [
+      "sts:AssumeRole",
```
```suggestion
      "sts:AssumeRole",
      "sts:TagSession",
```

mcdonnnj commented on outputs.tf:
```diff
@@ -1,14 +1,219 @@
+  description = "The security group for the Teamserver instances."
+}
+
+output "terraformer_instances" {
```
Just wanted to verify that there's no `terraformer_instance_profile` output by design.

jsf9k replied:
It is by design.  That's a powerful instance profile, and I thought it best if it wasn't dangling out there tempting folks to make use of it.

mcdonnnj commented on terraformer_iam.tf:
```diff
@@ -0,0 +1,65 @@
+resource "aws_iam_role_policy" "terraformer_assume_delegated_role_policy" {
+  provider = aws.provisionassessment
+
+  name   = "assume_delegated_role_policy"
```
Should this name include "terraformer" to be specific?

jsf9k replied:
Good call.  Please see commit 65663d2.

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -54,70 +109,89 @@ resource "aws_network_acl_rule" "operations_ingress_from_anywhere_via_allowed_po
-# For: Assessment team operational use, but don't want to allow public
-# access to Cobalt Strike teamserver on port 50050
+# For: Assessment team operational use, but we don't want to allow
+# public access to Cobalt Strike teamserver on port 50050.
```
```suggestion
# public access to Cobalt Strike Teamservers on port 50050.
```

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -54,70 +109,89 @@ resource "aws_network_acl_rule" "operations_ingress_from_anywhere_via_allowed_po
+# Allow ingress from anywhere via ephemeral TCP/UDP ports 5987-50049.
+#
+# For: Assessment team operational use, but we don't want to allow
+# public access to WinRM on port 5986 or Cobalt Strike teamserver on
```
```suggestion
# public access to WinRM on port 5986 or Cobalt Strike Teamservers on
```

mcdonnnj commented on operations_acl_rules.tf:
```diff
@@ -15,6 +17,25 @@ resource "aws_network_acl_rule" "operations_ingress_from_private_via_ssh" {
+
 # Allow ingress from private subnet via VNC
 #
 # For: Assessment team VNC access from private subnet to operations
```
```suggestion
# For: Assessment team VNC access from private subnet to operations.
```

mcdonnnj commented on examples/use-terraformer-instance/README.md:
```diff
@@ -0,0 +1,80 @@
+
+## Requirements ##
+
+No requirements.
```
Are we intentionally leaving out a `versions.tf` file for this example?

jsf9k replied:
Please see commit a32d80f.

mcdonnnj commented on examples/use-terraformer-instance/README.md:
```diff
@@ -0,0 +1,80 @@
+| aws\_region | The AWS region where the non-global resources for this assessment are to be provisioned (e.g. "us-east-1"). | `string` | `"us-east-1"` | no |
+| dns\_ttl | The TTL value to use for Route53 DNS records (e.g. 86400).  A smaller value may be useful when the DNS records are changing often, for example when testing. | `number` | `60` | no |
+| email\_sending\_domain | The domain to send emails from within the assessment environment (e.g. "example.com"). | `string` | `"example.com"` | no |
+| tags | Tags to apply to all AWS resources created | `map(string)` | `{}` | no |
```
```suggestion
| tags | Tags to apply to all AWS resources created. | `map(string)` | `{}` | no |
```

jsf9k replied:
Please see commit 42449436df2217a1e51fae68e2ff686eabc3a1d8.

mcdonnnj commented on examples/use-terraformer-instance/README.md:
```diff
@@ -0,0 +1,80 @@
+| teamserver\_a\_record | The teamserver A record. |
+| teamserver\_eip | The teamserver EIP. |
+
+## Notes ##
+
+Running `pre-commit` requires running `terraform init` in every directory that
+contains Terraform code. In this repository, these are the main directory and
+every directory under `examples/`.
+
+## Contributing ##
+
+We welcome contributions!  Please see [`CONTRIBUTING.md`](CONTRIBUTING.md) for
+details.
+
+## License ##
+
+This project is in the worldwide [public domain](LICENSE).
+
+This project is in the public domain within the United States, and
+copyright and related rights in the work worldwide are waived through
+the [CC0 1.0 Universal public domain
+dedication](https://creativecommons.org/publicdomain/zero/1.0/).
+
+All contributions to this project will be released under the CC0
+dedication. By submitting a pull request, you are agreeing to comply
+with this waiver of copyright interest.
```
Should this content just be removed since this is an example in a parent repository?

jsf9k replied:
Please see commit e595ead.

mcdonnnj commented on examples/use-terraformer-instance/outputs.tf:
```diff
@@ -0,0 +1,14 @@
+output "teamserver" {
+  value       = aws_instance.teamserver
+  description = "The teamserver instance."
```
```suggestion
  description = "The Teamserver instance."
```

mcdonnnj commented on examples/use-terraformer-instance/outputs.tf:
```diff
@@ -0,0 +1,14 @@
+
+output "teamserver_a_record" {
+  value       = aws_route53_record.teamserver_A
+  description = "The teamserver A record."
```
```suggestion
  description = "The Teamserver A record."
```

mcdonnnj commented on examples/use-terraformer-instance/outputs.tf:
```diff
@@ -0,0 +1,14 @@
+
+output "teamserver_eip" {
+  value       = aws_eip.teamserver
+  description = "The teamserver EIP."
```
```suggestion
  description = "The Teamserver EIP."
```

mcdonnnj commented on examples/use-terraformer-instance/teamserver_cloud_init.tf:
```diff
@@ -0,0 +1,85 @@
+    filename     = "01-install-certificates.py"
+  }
+
+  # Add https-certificate section to CobaltStrike profiles
```
```suggestion
  # Add https-certificate section to Cobalt Strike profiles
```

mcdonnnj commented on examples/use-terraformer-instance/teamserver_ec2.tf:
```diff
@@ -0,0 +1,79 @@
+  })
+}
+
+# The Elastic IP for the teamserver
```
```suggestion
# The Elastic IP for the Teamserver
```

mcdonnnj commented on examples/use-terraformer-instance/teamserver_ec2.tf:
```diff
@@ -0,0 +1,79 @@
+  }
+}
+
+# The EIP association for the teamserver
```
```suggestion
# The EIP association for the Teamserver
```

mcdonnnj commented on examples/use-terraformer-instance/teamserver_ec2.tf:
```diff
@@ -0,0 +1,79 @@
+  most_recent = true
+}
+
+# The teamserver EC2 instances
```
```suggestion
# The Teamserver EC2 instances
```

mcdonnnj commented on examples/use-terraformer-instance/teamserver_ec2.tf:
```diff
@@ -0,0 +1,79 @@
+# The teamserver AMI
```
```suggestion
# The Teamserver AMI
```

mcdonnnj commented on examples/use-terraformer-instance/variables.tf:
```diff
@@ -0,0 +1,29 @@
+
+variable "tags" {
+  type        = map(string)
+  description = "Tags to apply to all AWS resources created"
```
```suggestion
  description = "Tags to apply to all AWS resources created."
```

jsf9k replied:
Please see commit 42449436df2217a1e51fae68e2ff686eabc3a1d8.

mcdonnnj commented on examples/use-terraformer-instance/README.md:
```diff
@@ -0,0 +1,80 @@
+
+| Name | Description |
+|------|-------------|
+| teamserver | The teamserver instance. |
+| teamserver\_a\_record | The teamserver A record. |
+| teamserver\_eip | The teamserver EIP. |
```
```suggestion
| teamserver | The Teamserver instance. |
| teamserver\_a\_record | The Teamserver A record. |
| teamserver\_eip | The Teamserver EIP. |
```

mcdonnnj reviewed (approved):
Thank you for your work on getting this functionality built out. Some strong work 💪💪💪. I have a bevy of suggestions relating to capitalization and name consistency. There are a few other questions and suggestions smattered amongst these.

jsf9k committed:
Prefer long flags for increased clarity

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add a missing space in a comment

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct capitalization

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add missing period at the end of a description

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct capitalization

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove unnecessary mypy exclusion

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct capitalization in comment

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct capitalization in comment

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add missing period at end of comment

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct grammar in comments

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Capitalize Guacamole

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add TagSession permission to assume role policy

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add versions.tf to example code

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Remove parent project boilerplate

This language is not necessary in example code READMEs.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct output name

This makes the output name agree with those of the other assessor portal outputs.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Correct output names

The new names better agree with what is done, e.g., for the outputs related to the Assessor Portal instances.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Add a comment to explain complicated for_each construction

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Modify policy name to mention terraformer

This specificity makes the name infinitely more useful, particularly at a glance.

Co-authored-by: Nick M <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
@mcdonnnj can you circle back and revisit the unresolved comments when you have a chance?  I believe I have taken care of all of them.

jsf9k commented:
> @mcdonnnj can you circle back and revisit the unresolved comments when you have a chance? I believe I have taken care of all of them.

Thank you good sir!

jsf9k committed:
Merge remote-tracking branch 'origin/lineage/skeleton' into improvement/add-terraformer-instance

jsf9k committed:
Update example code to use Terraform 0.14

jsf9k committed:
Rework list comprehension to be a map comprehension

Terraform requires a map or a list of strings for for_each, so this is
necessary.

jsf9k committed:
Correct port number in local variable

jsf9k commented:
I will wait until #137 is merged before merging this PR.

jsf9k committed:
Allow the operations subnet to access a single Mattermost port

* Ensure that Mattermost ports that are in the ephemeral port range
are blocked from external access.  This is done by adding explicit
deny ACL rules, which since ACL rules are evaluated sequentially until
one applies, allows us to avoid "skipping" ports such as 2049 when
allowing ephemeral port ingress further down the rule set.
* Add some comments to make it easier to see how ACL rules are
organized.

jsf9k committed:
Group all ingress rules together

jsf9k committed:
Add explanatory comment

jsf9k committed:
Make the IPA ingress rules work for both private subnets

dav3r commented on examples/use-terraformer-instance/README.md:
```diff
@@ -0,0 +1,60 @@
+
+| Name | Version |
+|------|---------|
+| terraform | ~> 0.13.0 |
```
To reflect the change in https://github.com/cisagov/cool-assessment-terraform/pull/132/commits/54f4e06f322ebb450a768a9accc5480f2025309f:
```suggestion
| terraform | ~> 0.14.0 |
```

jsf9k committed:
Update Terraform version in example code README

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r reviewed (approved):
Everything looks groovy, baby, yeah! 🇬🇧

jsf9k committed:
Merge branch 'develop' into improvement/add-terraformer-instance

mcdonnnj commented on locals.tf:
```diff
@@ -152,6 +152,35 @@ locals {
 
+  # Ports to be accessed via the VPN in assessment environments
+  # (e.g. for Guacamole, Mattermost, etc.)
+  assessment_env_service_ports = {
```
So while re-reviewing I noticed that `ipa_ports` (above) uses `proto` but this map uses `protocol`. I think we should standardize against one or the other to keep things consistent (`ipa_ports` also doesn't have sorted keys but that's a thing for another time).

jsf9k replied:
See commit e0cd6a2.

mcdonnnj reviewed (approved):
Still looking good ✨. I had one question/suggestion around consistent naming conventions.

jsf9k committed:
Rename map key from "proto" to "protocol" for consistency
